Rebase mri branch from docs deletion branch (#128)
rstudio/ai-blog: _posts/2020-11-26-torch-brain-segmentation/torch_brain_segmentation.Rmd

---
title: "Brain image segmentation with torch"
description: > 

>   The need to segment images arises in various sciences and their applications, many of whom are vital to human (and animal) life. In this introductory post, we train a U-Net to mark lesioned regions on MRI brain scans.

Reviewer: ```suggestion
  The need to segment images arises in various sciences and their applications, many of which are vital to human (and animal) life. In this introductory post, we train a U-Net to mark lesioned regions on MRI brain scans.
```

Reviewer: Since you're referring back to sciences and their applications and not a person 👍

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydanatorchtabular
date: 11-26-2020
categories:
  - Torch
  - R
  - Image Recognition & Image Processing
bibliography: bibliography.bib
output:
  distill::distill_article:
    self_contained: false
    toc: true
preview: images/scans.png
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

## When *what* is not enough

> True, sometimes it's vital to distinguish between different kinds of objects. Is that a car speeding towards me, in which case I'd better jump out of the way? Or is it a huge terrier (in which case I'd probably do the same)? Often in real life though, instead of coarse-grained *classification*, what is needed is fine-grained *segmentation*.

Reviewer: It feels weird to me to start with True, here. I think you can just cut it. I started looking around for a question or a something I had missed, since it somehow implies that it's in response to something.

Reviewer: Do you mean terrier like the dog? If so, maybe just say dog—terriers are pretty little (generally speaking—pit bulls are technically American Staffordshire Terriers, but few people actually know that).

Reviewer 2: (((like we said on slack)))

Zooming in on images, we're not looking for a single label; instead, we want to classify every pixel according to some criterion:

-   In medicine, we may want to distinguish between different cell types, or identify tumors.

-   In various earth sciences, satellite data are used to segment terrestrial surfaces.

-   To enable use of custom backgrounds, video-conferencing software has to be able to tell foreground from background.

> Image segmentation is a form of supervised learning: Some kind of ground truth is needed. Here, it comes in form of a *mask* -- an image, of spatial resolution identical to that of the input data, that designates the true class for every pixel. Accordingly, classification loss is calculated pixel-wise; losses are then summed up to yield an aggregate to be used in optimization.

Reviewer: ```suggestion
Image segmentation is a form of supervised learning: Some kind of ground truth is needed. Here, it comes in form of a *mask* -- an image of spatial resolution identical to that of the input data, that designates the true class for every pixel. Accordingly, classification loss is calculated pixel-wise; losses are then summed up to yield an aggregate to be used in optimization.
```

Reviewer: I removed a comma (not sure how easy it is to see in the diff)

> The "canonical" architecture for image segmentation, around since 2015, is *U-Net.*

Reviewer: ```suggestion
The "canonical" architecture for image segmentation is *U-Net* (around since 2015).
```

## U-Net

Here is the prototypical U-Net, as depicted in the original Rönneberger et al. paper [@RonnebergerFB15].

Of this architecture, numerous variants exist; you could use different layer sizes, activations, ways to achieve downsizing and upsizing, and more. However, there is one defining characteristic: The U-shape, stabilized by the "bridges" crossing over horizontally at all levels.

> 

Reviewer: You've got two atypical sentence constructs in this paragraph (semicolon use is shockingly rare in American English, and the colon followed by an uppercase is also pretty uncommon). I'd ax one of them.

Reviewer 2: Will do!

Reviewer 2: just noticing, the uppercase shouldn't be there anyway ...

![](images/unet.png "The original U-Net, as depicted in Ronnerberger et al. (2015).")

> In a nutshell, the left-hand side of the U resembles the convolutional architectures used in image classification: It successively reduces spatial resolution. At the same time, another dimension -- the *channels* dimension -- is used to build up a hierarchy of features, ranging from very basic to very specialized.

Reviewer: ```suggestion
In a nutshell, the left-hand side of the U resembles the convolutional architectures used in image classification. It successively reduces spatial resolution. At the same time, another dimension -- the *channels* dimension -- is used to build up a hierarchy of features, ranging from very basic to very specialized.
```

> Unlike in classification, however, the output should have the same spatial resolution as the input. Thus, we need to upsize again -- this is taken care of by the right-hand side of the U. But: How are we going to arrive at a good *per-pixel* classification, now that so much spatial information has been lost?

Reviewer: ```suggestion
Unlike in classification, however, the output should have the same spatial resolution as the input. Thus, we need to upsize again -- this is taken care of by the right-hand side of the U. But, how are we going to arrive at a good *per-pixel* classification, now that so much spatial information has been lost?
```

> This is what the "bridges" are for: At each level, the input to an upsampling layer is a *concatenation* of the previous layer's output -- which went through the whole compression/decompression routine -- and some preserved intermediate representation from the downsizing phase. In this way, a U-Net architecture combines attention to detail with feature extraction.

Reviewer: I like the last sentence here! It sums up the utility of U-Net nicely.

## Brain image segmentation

> With U-Net, domain applicability is as broad as the architecture is flexible. Here, we want to detect abnormalities in brain scans. The dataset, used in @BUDA2019218, contains MR images together with manually created [FLAIR](https://en.wikipedia.org/wiki/Fluid-attenuated_inversion_recovery) abnormality segmentation masks. It is available on [Kaggle](https://www.kaggle.com/mateuszbuda/lgg-mri-segmentation).

Reviewer: I don't read a lot about MRIs, but I've never come across MR images (with just the magnetic resonance part abbreviated) before. Given you've got FLAIR so close by, maybe just spell it out?

Reviewer 2: will do!

Nicely, the paper is accompanied by a [GitHub repository](https://github.com/mateuszbuda/brain-segmentation-pytorch). Below, we closely follow (though not exactly replicate) the authors' preprocessing and data augmentation code.

As is often the case in medical imaging, there is notable class imbalance in the data. For every patient, sections have been taken at multiple positions. (Number of sections per patient varies.) Most sections do not exhibit any lesions; the corresponding masks are colored black everywhere.

Here are three examples where the masks *do* indicate abnormalities:

![](images/scans.png)

Let's see if we can build a U-Net that generates such masks for us.

## Data

Before you start typing, here is a [Colaboratory notebook](https://colab.research.google.com/drive/1gcPOlF4jff42yiquRbbCxnRANEqPpGMD?usp=sharing) to conveniently follow along.

> We use `pins` to obtain the data. Please see [this introduction](https://pins.rstudio.com/articles/boards-kaggle.html) if you haven't been using that package before.

Reviewer: ```suggestion
We use `pins` to obtain the data. Please see [this introduction](https://pins.rstudio.com/articles/boards-kaggle.html) if you haven't used that package before.
```

```{r}
library(torch)
library(torchvision)
library(tidyverse)
library(magick)
library(zeallot)
library(cowplot)
library(zip)

> library(pins)

Reviewer: Totally your choice, but when there are this many libraries, I find it helpful when the authors comment what each is being used for (especially when the names are non-obvious). If it's hard to sum up, no big deal—just a thought.

Reviewer 2: makes sense!

torch_manual_seed(777)
set.seed(777)

# use your own kaggle.json here
pins::board_register_kaggle(token = "~/kaggle.json")

files <- pins::pin_get("mateuszbuda/lgg-mri-segmentation", board = "kaggle",  extract = FALSE)

```

> The dataset is not that big -- it includes scans from 110 different patients --, so we'll have to do with just a training and a validation set. (Don't do this in real life, as you'll inevitably end up fine-tuning on the latter.)

Reviewer: ```suggestion
The dataset is not that big -- it includes scans from 110 different patients -- so we'll have to do with just a training and a validation set. (Don't do this in real life, as you'll inevitably end up fine-tuning on the latter.)
```

Reviewer: Rarely (if ever) do you need a combination of offsetting punctuation (the em-dash-comma combo here).

```{r}
train_dir <- "data/mri_train"
valid_dir <- "data/mri_valid"

if(dir.exists(train_dir)) unlink(train_dir, recursive = TRUE, force = TRUE)
if(dir.exists(valid_dir)) unlink(valid_dir, recursive = TRUE, force = TRUE)

zip::unzip(files, exdir = "data")

file.rename("data/kaggle_3m", train_dir)

# this is a duplicate, again containing kaggle_3m (evidently a packaging error on Kaggle)
# we just remove it
unlink("data/lgg-mri-segmentation", recursive = TRUE)

dir.create(valid_dir)

```

> Of those 110 patients, we keep 30 for validation. Some more file manipulations, and we're set up with a nice hierarchical structure, with each of `train_dir` and `valid_dir` holding their per-patient sub-directories, respectively.

Reviewer: This is that tricky "each" thing again, where you treat the subject as singular since you're referring to them one at a time.

> Of those 110 patients, we keep 30 for validation. Some more file manipulations, and we're set up with a nice hierarchical structure, with each of `train_dir` and `valid_dir` holding their per-patient sub-directories, respectively.

Reviewer: ```suggestion
Of those 110 patients, we keep 30 for validation. Some more file manipulations, and we're set up with a nice hierarchical structure, with each of `train_dir` and `valid_dir` holding its per-patient sub-directories, respectively.
```

Reviewer: Alternatively, you could remove the each, so it would read something like:
> with `train_dir` and `valid_dir` holding their per-patient sub-directories, respectively.

Reviewer 2: oh, that one's great, will do

```{r}
train_ds <- brainseg_dataset(
  train_dir,
  augmentation_params = c(0.05, 15, 0.5),
  random_sampling = TRUE
)

length(train_ds)
# 2977

valid_ds <- brainseg_dataset(
  valid_dir,
  augmentation_params = NULL,
  random_sampling = FALSE
)

length(valid_ds)
# 952
```

As a correctness check, let's plot an image and associated mask:

```{r}
par(mfrow = c(1, 2), mar = c(0, 1, 0, 1))

img_and_mask <- valid_ds[27]
img <- img_and_mask[[1]]
mask <- img_and_mask[[2]]

img$permute(c(2, 3, 1)) %>% as.array() %>% as.raster() %>% plot()
mask$squeeze() %>% as.array() %>% as.raster() %>% plot()
```

```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
knitr::include_graphics("images/scan_and_mask.png")
```

With `torch`, it is straightforward to inspect what happens when you change augmentation-related parameters. We just pick a pair from the validation set, which has not had any augmentation applied as yet, and call `valid_ds$<augmentation_func()>` directly. Just for fun, let's use more "extreme" parameters here than we do in actual training. (Actual training uses the settings from Mateusz' GitHub repository, which we assume have been carefully chosen for optimal performance.[^1])

> [^1]: Yes, we did a few experiments, confirming that more augmentation isn't better ... what did I say about inevitably ending up doing optimization on the validation set ...

Reviewer: I think you want a question mark at the end of the question rather than a second ellipsis.

Reviewer 2: ok!

  
  # scale factor; train_ds really uses 0.05
  c(img, mask) %<-% valid_ds$resize(img, mask, 0.2) 
  c(img, mask) %<-% valid_ds$flip(img, mask, 0.5)
  # rotation angle; train_ds really uses 15
  c(img, mask) %<-% valid_ds$rotate(img, mask, 90) 
  img %>%
    transform_rgb_to_grayscale() %>%
    as.array() %>%
    as_tibble() %>%
    rowid_to_column(var = "Y") %>%
    gather(key = "X", value = "value", -Y) %>%
    mutate(X = as.numeric(gsub("V", "", X))) %>%
    ggplot(aes(X, Y, fill = value)) +
    geom_raster() +
    theme_void() +
    theme(legend.position = "none") +
    theme(aspect.ratio = 1)
  
})

plot_grid(plotlist = imgs, nrow = 4)
```

```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
knitr::include_graphics("images/brainseg_augmentation.png")
```

Now we still need the data loaders, and then, nothing keeps us from proceeding to the next big task: building the model.

```{r}
batch_size <- 4
train_dl <- dataloader(train_ds, batch_size)
valid_dl <- dataloader(valid_ds, batch_size)

```

## Model

Our model nicely illustrates the kind of modular code that comes "naturally" with `torch`. We approach things top-down, starting with the U-Net container itself.

> `unet` takes care of the global composition -- how far "down" do we go, shrinking the image while incrementing the number of filters, and then how do we go "up" again?

Reviewer: Are you referring to the function you create below? If so, `unet()` gets trailing parentheses.

Reviewer: Oh, maybe it's an object you're creating, and not a function (in which case, ignore my comment above).

Reviewer 2: yeah, i know ... i was thinking about that ... technically it's a constructor, so more like a function, but the semantics should really be noun-like ...

> Importantly, it is also the system's memory: In `forward()`, it keeps track of layer outputs seen going "down", to be added back in going "up".

Reviewer: ```suggestion
Importantly, it is also in the system's memory. In `forward()`, it keeps track of layer outputs seen going "down", to be added back in going "up".
```

                                  out_size,
                                  kernel_size = 2,
                                  stride = 2)
    self$conv_block = conv_block(in_size, out_size)
  },
  
  forward = function(x, bridge) {
    
    up <- self$up(x)
    torch_cat(list(up, bridge), 2) %>%
      self$conv_block()
  }
)

```

Finally, a `conv_block` is a sequential structure containing convolutional, ReLU, and dropout layers.

```{r}
conv_block <- nn_module( 
  "conv_block",
  
  initialize = function(in_size, out_size) {
    
    self$conv_block <- nn_sequential(
      nn_conv2d(in_size, out_size, kernel_size = 3, padding = 1),
      nn_relu(),
      nn_dropout(0.6),
      nn_conv2d(out_size, out_size, kernel_size = 3, padding = 1),
      nn_relu()
    )
  },
  
  forward = function(x){
    self$conv_block(x)
  }
)
```

Now instantiate the model, and possibly, move it to the GPU:

```{r}
device <- torch_device(if(cuda_is_available()) "cuda" else "cpu")
model <- unet(depth = 5)$to(device = device)
```

## Optimization

We train our model with a combination of cross entropy and [dice loss](https://en.wikipedia.org/wiki/S%C3%B8rensen%E2%80%93Dice_coefficient).

> The latter, though not shipped with `torch`, may straightforwardly be implemented manually:

Reviewer: ```suggestion
The latter, though not shipped with `torch`, may be implemented manually:
```
